core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java

```diff
@@ -86,28 +95,44 @@ public String getName()
             return immediateVoidFuture();
         }
 
-        Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
-        if (tableHandle.isEmpty()) {
+        RedirectionAwareTableHandle redirectionAwareTableHandle = metadata.getRedirectionAwareTableHandle(session, tableName);
+        if (redirectionAwareTableHandle.getTableHandle().isEmpty()) {
             if (!statement.isExists()) {
                 throw semanticException(TABLE_NOT_FOUND, statement, "Table '%s' does not exist", tableName);
             }
             return immediateVoidFuture();
         }
-
-        QualifiedObjectName target = createQualifiedObjectName(session, statement, statement.getTarget());
+        TableHandle tableHandle = redirectionAwareTableHandle.getTableHandle().get();
+        QualifiedObjectName source = redirectionAwareTableHandle.getRedirectedTableName().orElse(createQualifiedObjectName(session, statement, statement.getSource()));
+        QualifiedObjectName target = createTargetQualifiedObjectName(source, statement.getTarget());
```

Comment:
This is a change in behavior that spans non-redirected use-cases.
For example
```
USE hive.default;
ALTER TABLE hive.tmp.my_table RENAME TO my_renamed;
```

before the change, the table was being renamed to `hive.default.my_renamed`.
after the change, the table is `hive.tmp.my_renamed`.
please correct me if am wrong.

I wasn't aware of this behavior. and this is why i wanted RENAME to work with redirects.
I actually think that the old behavior is misleading, hence https://github.com/trinodb/trino/issues/11281.
The changes here seem to resolve that problem, so

- let's make sure it's intentional (ie we agree https://github.com/trinodb/trino/issues/11281 is a bug)
- let's make this as a separate commit (or even PR)

Reply:
I'll create a new PR for this purpose.